### DISCUSSION TITLE
DI Scopes and Test using ClassConstructorAttribute ?

### DISCUSSION THREAD

**Discussion by Simonl9l** [thomhurst/TUnit#1269] [Q&A]
Hi All - love the lib....

Setting up integration tests that are heavily dependent on DI and leveraging the [ClassConstructors](https://thomhurst.github.io/TUnit/docs/tutorial-extras/class-constructors) as documented with DotNet DI Service Provider. 

However I really need to be able to DI on a "by test" basis, not by class, as I need the services to all be isolated between tests.

I was figuring I probably need to use the [Property Injection](https://thomhurst.github.io/TUnit/docs/tutorial-extras/property-injection) sync that I can control the scope, and possibly a [ClassDataSource](https://thomhurst.github.io/TUnit/docs/tutorial-basics/class-data-source) but can figure how to hook this up. 

Is this in fact possible, or what other approaches do you recommend?

**Comment by thomhurst** [comment]
You get a new instance of a class for each test, so class constructors should still work

**Comment by Simonl9l** [comment]
@thomhurst thanks for following up! 

Yes you are correct, each test it seems to fire up a new instance of the Test Class, and thus DI's new services and they have their own `AsyncServiceScope`, however in the sample code in the docs per [here](https://thomhurst.github.io/TUnit/docs/tutorial-extras/class-constructors), it sets up the service provider as a class static instance:
``` C#
public class DependencyInjectionClassConstructor : IClassConstructor, ITestEndEvent
{
    private static readonly IServiceProvider _serviceProvider = CreateServiceProvider();
    ...
}
```
This suggests (given my dotnet/C# understanding) that the `_serviceProvider` will be shared across _all_ instances of the `DependencyInjectionClassConstructor`, and this for each instance of the Test Class it's attributed to, and that is probably not desirable. 

I noted at runtime, running tests in parallel, that singletons that maintained `Dictionary<TK, TV>`s were running in to unexpected duplicate key issues between tests.

By reverting the code to to make the `_serviceProvider` not `static`:
``` C#
public class DependencyInjectionClassConstructor : IClassConstructor, ITestEndEvent
{
    private readonly IServiceProvider _serviceProvider = CreateServiceProvider();
    ...
}
```
Seems to solve the issue!

It may be that the documentation needs updating?

Thanks again for putting this package together.

**Reply by thomhurst** [reply]
> It may be that the documentation needs updating?

Setting up the DI is up to you, I just provide an example to explain the attribute.

It should still be able to be static if you ever wanted to share something between tests.

Then you just register services with different lifetimes:
- Singleton > Shared between tests
- Scoped/Transient > new'd up for each test
